docs(NODE-4916, NODE-4523): update readme and upgrade guide with callback removal information (#3540)
mongodb/node-mongodb-native: etc/notes/CHANGES_5.0.0.md

### Optional callback support migrated to `mongodb-legacy`

Node v5 drops support for callbacks in favor of a pure async-await API.  Below are some strategies for
callback users to adopt driver v5 in recommendation order.

> ### Migrate to promise based api (recommended!)

Reviewer: Should this and the other subheadings in this section use the `####` instead of `###`?

> The Node team strongly encourages anyone who is able to migrate from callbacks to promises to do so.  The

Reviewer: I think we should mention *why* that's a better experience for users (and/or mention why we are doing async/await in the heading above: better types, better stack traces, etc)

promise-based API is identical to the callback API except:

- no callback is accepted as the last argument
- a promise is always returned

For example, with a findOne query:

```typescript
// callback-based API
collection.findOne({ name: 'john snow'}, (error, result) => {
  if (error) {
    doSomethingWithError(error);
    return;
  }

  doSomethingWithResult(result);
})

// promise-based API
collection.findOne({ name: 'john snow'}).then(
  (result) => doSomethingWithResult(result),
  (error) => doSomethingWithError(error)
);

// promise-based API with async/await
try {
  const result = await collection.findOne({ name: 'john snow'});
  doSomethingWithResult(result);
} catch (error) {
  doSomethingWithError(error);
}
```

### Use the promise-based API and `util.callbackify`

> If you are unable to adopt the promise API fully, we recommend using the promise API and Nodejs's `callbackify`

Reviewer: ```suggestion
If you only have a few callback instances where you are currently unable to adopt the promise API, we recommend using the promise API and Nodejs's `callbackify`
```

### Optional callback support migrated to `mongodb-legacy`

> Node v5 drops support for callbacks in favor of a pure async-await API.  Below are some strategies for

Reviewer 2: ```suggestion
Node v5 drops support for callbacks in favor of an async-await API.  Below are some strategies for
```

I don't think this has general meaning externally, its terminology we invented to differentiate using the syntax and making promises ourselves. It might also be more accurate to change "favor of an async-await API" to "favor of a Promise only API"

> callback users to adopt driver v5 in recommendation order.

Reviewer 2: ```suggestion
callback users to adopt driver v5 in most to least recommended order.
```

This is optional, the phrasing just caught me up on first read.

The Node driver team understands that a callback to promise migration can be a non-trivial refactor.  To help inform your migration strategy, we've outlined three different approaches below.

#### Migrate to promise based api (recommended!)

The Node team strongly encourages anyone who is able to migrate from callbacks to promises to do so.  Adopting the
regular driver API will streamline the adoption of future driver updates, as well as provide better Typescript support
than the other options outlined in this document.

The promise-based API is identical to the callback API except:

- no callback is accepted as the last argument
- a promise is always returned

For example, with a findOne query:

```typescript
// callback-based API

> collection.findOne({ name: 'john snow'}, (error, result) => {

Reviewer 2: Can you run prettier on this readme

  if (error) {
    doSomethingWithError(error);
    return;
  }

  doSomethingWithResult(result);
})

// promise-based API
collection.findOne({ name: 'john snow'}).then(
  (result) => doSomethingWithResult(result),
  (error) => doSomethingWithError(error)

> );

Reviewer 2: Do we want to be recommending this manual promise wrapping? If `async` `await` isn't available, `.then`/`.catch` is what _most_ people should use attaching to an existing promise chain (possibly by just returning it: `return c.findOne()`) 

If folks need to inline this code inside of a place that is callback based they should prefer callbackify because of how it correctly does the error handling for promise rejection

Author: ```suggestion
// promise-based API
collection.findOne({ name: 'john snow'})
  .then((result) => doSomethingWithResult(result))
  .catch((error) => doSomethingWithError(error));
```

is this what you're suggesting?

Reviewer 2: Yep! 

Maybe take it a step further and don't invoke handlers: `doSomethingWithResult` and instead leave comments: `/* Do something with result */`. We know how mixing the two paradigms gets tricky so I don't want to have code samples that suggest a way that could be error prone, but I don't feel very strongly at least if we're demonstrating usage with the typical `.catch` handling.

// promise-based API with async/await
try {
  const result = await collection.findOne({ name: 'john snow'});
  doSomethingWithResult(result);
} catch (error) {
  doSomethingWithError(error);
}
```

#### Use the promise-based API and `util.callbackify`

If you only have a few callback instances where you are currently unable to adopt the promise API, we recommend using the promise API and Nodejs's `callbackify`
utility to adapt the promise-based API to use callbacks.

> **Note** Manually converting a promise-based api to a callback-based API is error prone.  We strongly encourage the user of `callbackify`.

Reviewer 3: ```suggestion
**Note** Manually converting a promise-based api to a callback-based API is error prone.  We strongly encourage the use of `callbackify`.
```

There are two approaches to using `callbackify` with a driver api.

> First, the driver method can be callbackified.  If this approach is used, care must be taken to bind the

Reviewer 2: I think we should make the anonymous function approach first, and potentially consider dropping the bind one. The anonymous one allows folks to continue omitting the options argument, when you use bind I think you need to provide options to make the callback the third argument, which I don't think is worth explaining.

Author: I'm fine removing the `bind` approach.  Although, it does preserve optional arguments.

The package wraps all of the driver's asynchronous operations that previously supported both promises and callbacks. All the wrapped APIs offer callback support via an optional callback argument alongside a Promise return value so projects with mixed usage will continue to work.

> `mongodb-legacy` is intended to preserve driver v4 behavior to enable an smoother transition between

Reviewer: ```suggestion
`mongodb-legacy` is intended to preserve driver v4 behavior to enable a smoother transition between
```
